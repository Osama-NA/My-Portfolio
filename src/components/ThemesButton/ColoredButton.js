import React from 'react'
import {  useContext } from 'react';
import { ThemesContext } from '../../contexts/ThemesContext';

// When clicked changes layout theme colors
export const ColoredButton = ({ thisTheme }) => {

    const transparentBorder = { border: "none" };
    const selectedBorder = { border: "1.6px solid " + thisTheme.fontColors.primary };

    const { theme, setTheme } = useContext(ThemesContext);

    const changeTheme = (newTheme) => setTheme(newTheme);

    return (
        <div
            style={theme === thisTheme ? selectedBorder : transparentBorder}
            className="colored-button"
            onClick={() => changeTheme(thisTheme)}
            data-testid="theme"
        >
            <div
                className="selected"
                style={{
                    background: `linear-gradient(45deg,
                        ${thisTheme.themeColors.mainColor} 50%,
                        ${thisTheme.backgroundColors.lightSide} 50.1%
                    )`
                }}
            >
            </div>
        </div>
    )
}
