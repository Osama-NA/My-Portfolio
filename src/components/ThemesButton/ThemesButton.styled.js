import styled from 'styled-components';

export const ThemesButtonContainer = styled.div`
    position: fixed;
    top: 2rem;
    right: 3.5rem;
    display: flex;
    flex-direction: column;
    align-items: flex-end;
    z-index: 1;

    .button{
        position: relative;
        cursor: pointer;
        color: ${({ theme }) => theme.fontColors.primary};
        font-size: 22px;
        border-radius: 8px;
        padding: .65rem;
        transition: all .2s ease;
        background-color: ${({ theme }) => theme.backgroundColors.lightBackground};
        box-shadow: -1px 1px 3px rgb(0 0 0 / 30%);
    }
    .button::after{
        content: '';
        position: absolute;
        bottom: 0;
        right: 0;
        width: 100%;
        height: 100%;
    }

      //Used to animate message box from visible to hidden going to the right side
    @keyframes animateBox {
        0% {
        opacity: 0;
        margin-right: -30rem;
        visibility: hidden;
        }
        100% {
        opacity: 1;
        margin-right: 0rem;
        visibility: visible;
        }
    }

    //Used to animate message box from hidden to visible coming from the right side
    @keyframes animateBoxOut {
        0% {
            opacity: 1;
            margin-right: 0rem;
            visibility: visible;
        }
        100% {
            opacity: 0;
            margin-right: -30rem;
            visibility: hidden;
        }
    }

    .themes-container{
        margin-top: 2rem;
        border-radius: 10px;
        background-color: ${({ theme }) => theme.backgroundColors.lightBackground};
        animation-duration: 0.6s;
        animation-fill-mode: forwards;
        box-shadow: -1px 1px 3px rgb(0 0 0 / 30%);
        visibility: hidden;

        .heading{
            display: flex;
            flex-direction: row;
            justify-content: space-between;
            align-items: center;
            padding: .8rem 1.2rem .8rem;
            border-bottom: 1.6px solid ${({ theme }) => theme.backgroundColors.lightSide};

            p{
                color: ${({ theme }) => theme.fontColors.primary};
                font-size: 17px;
                font-weight: 200;
                text-shadow: -1px 1px 3px rgb(0 0 0 / 40%);
            }
            i{
                cursor: pointer;
                font-size: 17px;
                color: ${({ theme }) => theme.themeColors.mainColor};
                transition: all 0.1s ease;
            }
        }
        .themes{
            width: 222px;
            display: flex;
            flex-direction: row;
            flex-wrap: wrap;
            padding: 1rem;

            .colored-button{
                cursor: pointer;
                width: 41px;
                height: 41px;
                border-radius: 50%;
                display: flex;
                justify-content: center;
                align-items: center;
                margin: .2rem;

                .selected{
                    width: 25px;
                    height: 25px;
                    border-radius: 50%;
                    padding: .5rem;
                }
            }
        }
    }
    @media (min-width: 768px) and (max-width: 1200px){
        top: 1.5rem;
        right: 2rem;

        .button{
            font-size: 18px;
        }
        .themes-container{
            margin-top: 2rem;

            .heading{
                padding: .6rem 1rem .6rem;

                p{
                    font-size: 15px;
                }
                i{
                    font-size: 15px;
                }
            }
            .themes{
                width: 198px;
                padding: .75rem;

                .colored-button{
                    width: 37px;
                    height: 37px;

                    .selected{
                        width: 21px;
                        height: 21px;
                    }
                }
            }
        }
    }
    @media (max-width: 767px) {
        top: 1rem;
        right: 1.5rem;

        .button{
            font-size: 16px;
            padding: .5rem;
            border-radius: 6px;
        }

        .themes-container{
            margin-top: 1rem;
            border-radius: 8px;

            .heading{
                padding: .4rem .8rem;

                p{
                    font-size: 12px;
                    font-weight: 200;
                }
                i{
                    font-size: 12px;
                }
            }
            .themes{
                width: 155px;
                padding: .4rem;

                .colored-button{
                    width: 30px;
                    height: 30px;
                    margin: .15rem;

                    .selected{
                        width: 18px;
                        height: 18px;
                        padding: .4rem;
                    }
                }
            }
        }
    }
`