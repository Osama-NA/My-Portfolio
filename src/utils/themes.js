// GREEN / GRAY
export const greenGrayTheme = {
    themeColors: {
        mainColor: '#54EBA0'
    },
    fontColors: {
        secondary: '#e0e0e0',
        primary: '#fff',
        gray: '#5e5e5e'
    },
    backgroundColors: {
        lightSide: '#313138',
        darkSide: '#242429',
        lightBackground: '#3F3F47',
        navBackground: '#383840'
    },
    cubesColor: '#03AD78',
    circleBackground: '#313138'
}

// BLUEISH PURPLE / GRAY 
export const blueishPurpleGrayTheme = {
    themeColors: {
        mainColor: '#674DFF'
    },
    fontColors: {
        secondary: '#e0e0e0',
        primary: '#fff',
        gray: '#5e5e5e'
    },
    backgroundColors: {
        lightSide: '#313138',
        darkSide: '#242429',
        lightBackground: '#3F3F47',
        navBackground: '#383840'
    },
    cubesColor: '#544DFF',
    circleBackground: '#313138'
}

// PURPLE / GRAY
export const purpleGrayTheme = {
    themeColors: {
        mainColor: '#B15EFF'
    },
    fontColors: {
        secondary: '#e0e0e0',
        primary: '#fff',
        gray: '#5e5e5e'
    },
    backgroundColors: {
        lightSide: '#313138',
        darkSide: '#242429',
        lightBackground: '#3F3F47',
        navBackground: '#383840'
    },
    cubesColor: '#9A52DE',
    circleBackground: '#313138'
}

// LIGHT BLUE / GRAY
export const lightBlueGrayTheme = {
    themeColors: {
        mainColor: '#00E6E2'
    },
    fontColors: {
        secondary: '#e0e0e0',
        primary: '#fff',
        gray: '#5e5e5e'
    },
    backgroundColors: {
        lightSide: '#313138',
        darkSide: '#242429',
        lightBackground: '#3F3F47',
        navBackground: '#383840'
    },
    cubesColor: '#00E6E2',
    circleBackground: '#313138'
}

// YELLOW / GRAY
export const yellowGrayTheme = {
    themeColors: {
        mainColor: '#F2F270'
    },
    fontColors: {
        secondary: '#e0e0e0',
        primary: '#fff',
        gray: '#5e5e5e'
    },
    backgroundColors: {
        lightSide: '#313138',
        darkSide: '#242429',
        lightBackground: '#3F3F47',
        navBackground: '#383840'
    },
    cubesColor: '#F2F208',
    circleBackground: '#313138'
}

// PINK / GRAY
export const pinkGrayTheme = {
    themeColors: {
        mainColor: '#FF558E'
    },
    fontColors: {
        secondary: '#e0e0e0',
        primary: '#fff',
        gray: '#5e5e5e'
    },
    backgroundColors: {
        lightSide: '#313138',
        darkSide: '#242429',
        lightBackground: '#3F3F47',
        navBackground: '#383840'
    },
    cubesColor: '#FF558E',
    circleBackground: '#313138'
}

// RED / GRAY
export const redGrayTheme = {
    themeColors: {
        mainColor: '#FF3142'
    },
    fontColors: {
        secondary: '#e0e0e0',
        primary: '#fff',
        gray: '#5e5e5e'
    },
    backgroundColors: {
        lightSide: '#313138',
        darkSide: '#242429',
        lightBackground: '#3F3F47',
        navBackground: '#383840'
    },
    cubesColor: '#FF3142',
    circleBackground: '#313138'
}