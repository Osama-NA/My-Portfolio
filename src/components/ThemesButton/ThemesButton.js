import {ThemesButtonContainer} from './ThemesButton.styled';
import {useState} from 'react';
import {ColoredButton} from './ColoredButton';
import {lightPinkTheme} from '../../utils/themes';
import {greenPurpleTheme} from '../../utils/themes';
import {greenBlueBerryTheme} from '../../utils/themes';
import {pinkGrayTheme} from '../../utils/themes';
import {yellowDarkBlueTheme} from '../../utils/themes';
import {blueGrayTheme} from '../../utils/themes';
import { beigeGreenTheme} from '../../utils/themes';
import { purpleDarkBrownTheme} from '../../utils/themes';

const themes = [
    lightPinkTheme,
    beigeGreenTheme,
    blueGrayTheme,
    yellowDarkBlueTheme,
    greenPurpleTheme,
    greenBlueBerryTheme,
    pinkGrayTheme,
    purpleDarkBrownTheme
]

export const ThemesButton = () => {

    const [themeBoxStyle, setThemeBoxStyle] = useState({});

    const displayThemeBox = () => setThemeBoxStyle({ animationName: "animateBox" });

    const closeThemeBox = () => setThemeBoxStyle({ animationName: "animateBoxOut" });

    return(
        <ThemesButtonContainer className="themes-button" data-testid="themes-button-container">
            <i className="fas fa-cog button" onClick={displayThemeBox} data-testid="themes-button" /> {/* When clicked opens theme settings*/}

            {/* THEME SETTINGS CONTAINER */}
            <div className="themes-container" style={themeBoxStyle} data-testid="themes-container" >
                <div className="heading">
                    <p>Select a theme</p>
                    <i className="fas fa-times close-btn" onClick={closeThemeBox} />
                </div>
                <div className="themes" data-testid="themes">
                    {themes.map((theme, i) => {
                        return <ColoredButton key={i} thisTheme={theme}></ColoredButton>
                    })}
                </div>
            </div>
        </ThemesButtonContainer>
    )
}