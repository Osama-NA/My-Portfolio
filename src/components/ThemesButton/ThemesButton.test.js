import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import { ThemeProvider } from 'styled-components';
import { ThemesButton } from './ThemesButton.js';
import { ColoredButton } from './ColoredButton.js';
import { cleanup, render } from '@testing-library/react';
import { lightPinkTheme } from '../../utils/themes.js';

afterEach(cleanup);

describe("ThemesButton component tests", () => {

    const ThemesButtonComponent = () => {
        return (
            <ThemeProvider theme={lightPinkTheme}>
                <ThemesButton />
            </ThemeProvider>
        )
    }
    
    const ColoredButtonComponent = () => {
        return (
            <ThemeProvider theme={lightPinkTheme}>
                <ColoredButton thisTheme={lightPinkTheme} />
            </ThemeProvider>
        )
    }

    it("ThemesButton component renders without crashing", () => {
        let div = document.createElement('div');
        ReactDOM.render(<ThemesButtonComponent />, div);
    })

    it("ThemesButton is in the document", () => {
        const { getByTestId } = render(<ThemesButtonComponent />);
        expect(getByTestId('themes-button-container')).toBeInTheDocument();
    })

    it("ColoredButton component renders without crashing", () => {
        let div = document.createElement('div');
        ReactDOM.render(<ColoredButtonComponent />, div);
    })

    it("ColoredButton is in the document", () => {
        const { getByTestId } = render(<ColoredButtonComponent />);
        expect(getByTestId('theme')).toBeInTheDocument();
    })

    it("ThemesButton component contains themes button element", () => {
        const { getByTestId } = render(<ThemesButtonComponent />);
        expect(getByTestId('themes-button-container')).toContainElement(getByTestId('themes-button'));
    })

    it("ThemesButton component contains themes container", () => {
        const { getByTestId } = render(<ThemesButtonComponent />);
        expect(getByTestId('themes-button-container')).toContainElement(getByTestId('themes-container'));
    })

    it("Themes container contains themes", () => {
        const { getByTestId } = render(<ThemesButtonComponent />);
        expect(getByTestId('themes-container')).toContainElement(getByTestId('themes'));
    })

    it("ThemesButton component matches snapshot", () => {
        const tree = renderer.create(<ThemesButtonComponent />).toJSON();
        expect(tree).toMatchSnapshot();
    })
})