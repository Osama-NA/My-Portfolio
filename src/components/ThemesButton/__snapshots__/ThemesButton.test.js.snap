// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ThemesButton component tests ThemesButton component matches snapshot 1`] = `
<div
  className="sc-bdfBQB frVean themes-button"
  data-testid="themes-button-container"
>
  <i
    className="fas fa-cog button"
    data-testid="themes-button"
    onClick={[Function]}
  />
   
  <div
    className="themes-container"
    data-testid="themes-container"
    style={Object {}}
  >
    <div
      className="heading"
    >
      <p>
        Select a theme
      </p>
      <i
        className="fas fa-times close-btn"
        onClick={[Function]}
      />
    </div>
    <div
      className="themes"
      data-testid="themes"
    >
      <div
        className="colored-button"
        data-testid="theme"
        onClick={[Function]}
        style={
          Object {
            "border": "none",
          }
        }
      >
        <div
          className="selected"
          style={
            Object {
              "background": "linear-gradient(45deg,
                        #ffaaa5 50%,
                        #363062 50.1%
                    )",
            }
          }
        />
      </div>
      <div
        className="colored-button"
        data-testid="theme"
        onClick={[Function]}
        style={
          Object {
            "border": "none",
          }
        }
      >
        <div
          className="selected"
          style={
            Object {
              "background": "linear-gradient(45deg,
                        #FAB09B 50%,
                        #004D4D 50.1%
                    )",
            }
          }
        />
      </div>
      <div
        className="colored-button"
        data-testid="theme"
        onClick={[Function]}
        style={
          Object {
            "border": "none",
          }
        }
      >
        <div
          className="selected"
          style={
            Object {
              "background": "linear-gradient(45deg,
                        #657CF0 50%,
                        #36373D 50.1%
                    )",
            }
          }
        />
      </div>
      <div
        className="colored-button"
        data-testid="theme"
        onClick={[Function]}
        style={
          Object {
            "border": "none",
          }
        }
      >
        <div
          className="selected"
          style={
            Object {
              "background": "linear-gradient(45deg,
                        #F2F270 50%,
                        #2F3673 50.1%
                    )",
            }
          }
        />
      </div>
      <div
        className="colored-button"
        data-testid="theme"
        onClick={[Function]}
        style={
          Object {
            "border": "none",
          }
        }
      >
        <div
          className="selected"
          style={
            Object {
              "background": "linear-gradient(45deg,
                        #54EBA0 50%,
                        #491D4D 50.1%
                    )",
            }
          }
        />
      </div>
      <div
        className="colored-button"
        data-testid="theme"
        onClick={[Function]}
        style={
          Object {
            "border": "none",
          }
        }
      >
        <div
          className="selected"
          style={
            Object {
              "background": "linear-gradient(45deg,
                        #FF004B 50%,
                        #36373D 50.1%
                    )",
            }
          }
        />
      </div>
      <div
        className="colored-button"
        data-testid="theme"
        onClick={[Function]}
        style={
          Object {
            "border": "none",
          }
        }
      >
        <div
          className="selected"
          style={
            Object {
              "background": "linear-gradient(45deg,
                        #FF7A89 50%,
                        #313138 50.1%
                    )",
            }
          }
        />
      </div>
      <div
        className="colored-button"
        data-testid="theme"
        onClick={[Function]}
        style={
          Object {
            "border": "none",
          }
        }
      >
        <div
          className="selected"
          style={
            Object {
              "background": "linear-gradient(45deg,
                        #B15EFF 50%,
                        #453B3B 50.1%
                    )",
            }
          }
        />
      </div>
    </div>
  </div>
</div>
`;
